src/shared/utilities/timeoutUtils.ts

```diff
@@ -3,29 +3,43 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-export const TIMEOUT_ERROR_MESSAGE = 'Timeout token expired'
+export const TIMEOUT_EXPIRED_MESSAGE = 'Timeout token expired'
+export const TIMEOUT_CANCELLED_MESSAGE = 'Timeout token cancelled'
+export const TIMEOUT_UNEXPECTED_RESOLVE = 'Promise resolved with an unexpected object'
+export const TIMEOUT_BAD_REFRESH = 'Timeout token cannot be refreshed after fulfilled'
 
 /**
- * Timeout that can handle both cancellation token-style and time limit-style timeout situations.
+ * Timeout that can handle both cancellation token-style and time limit-style timeout situations. Timeouts
+ * cannot be used after 'complete' has been called or if the Timeout expired.
+ *
  * @param timeoutLength Length of timeout duration (in ms)
  */
 export class Timeout {
     private originalStartTime: number
     private startTime: number
     private endTime: number
     private readonly timeoutLength: number
-    private readonly timerPromise: Promise<void>
-    private timerTimeout?: NodeJS.Timeout
-    private timerResolve?: (value?: void | PromiseLike<void> | undefined) => void
+    private readonly timerPromise: Promise<undefined>
+    private timerTimeout: NodeJS.Timeout
+    private timerResolve!: (value?: Promise<undefined> | undefined) => void
+    private timerReject!: (value?: Error | Promise<Error> | undefined) => void
+    private _fulfilled: boolean = false
+
     public constructor(timeoutLength: number) {
         this.startTime = Date.now()
         this.originalStartTime = this.startTime
         this.endTime = this.startTime + timeoutLength
         this.timeoutLength = timeoutLength
-        this.timerPromise = new Promise<void>((resolve, reject) => {
-            this.timerTimeout = setTimeout(() => reject(new Error(TIMEOUT_ERROR_MESSAGE)), timeoutLength)
+
+        this.timerPromise = new Promise<undefined>((resolve, reject) => {
+            this.timerReject = reject
             this.timerResolve = resolve
         })
+
+        this.timerTimeout = setTimeout(() => {
+            this.timerReject(new Error(TIMEOUT_EXPIRED_MESSAGE))
+            this._fulfilled = true
```

Comment:
complete() can also set this. But complete() also changes other state. What should happen if caller invokes complete() after timerTimeout already ran?

Reply:
Ah true didn't see that scenario. `complete` should check for `_fulfilled` and do nothing if true.